src/lib/reif.pl

```diff
@@ -1,4 +1,10 @@
-/** Predicates from [*Indexing dif/2*](https://arxiv.org/abs/1607.01590).
+/** Reification is the process of treating the success and failure of a
+predicate as if it were a concrete value. By indexing success and failure to
+two distinct values, reification reduces the complexity of interfacing with
+impure code.
```

Comment:
As I see it, an important key advantage we get from this it that the resulting code is both correct and efficient: It always yields all solutions, and it is deterministic when possible.